c2rust-analyze/src/main.rs

```diff
@@ -474,6 +503,29 @@ fn run(tcx: TyCtxt) {
                 || info.contains(PointerInfo::NOT_TEMPORARY_REF))
     }
 
+    // track all types mentioned in extern blocks, we
+    // don't want to rewrite those
+    let mut foreign_mentioned_tys = HashSet::new();
```

Comment:
Could you move this block calculating `foreign_mentioned_tys` to its own `fn`?  `fn run` is big enough already.

Reply:
Also, once the other changes are applied, I think you could simplify it a lot more with some iterator methods:

```rust
for ty in tcx.hir_crate_items(())
    .foreign_items()
    .map(|item| item.def_id.to_def_id())
    .filter_map(|did| match tcx.def_kind(did) {
        DefKind::Fn | DefKind::AssocFn => Some(TyKind::FnPtr(tcx.fn_sig(did))),
        DefKind::Static(_) => Some(tcx.type_of(did)),
        _ => None,
    }) {
    walk_ffi_safe_adts(tcx, ty, &mut |did| foreign_mentioned_tys.insert(did));
}
```

Reply:
moved to it's own function in d0a6aac9. `tcx.type_of` returns a `Ty` and that's not a `TyKind` so your suggestion will not work.

Reply:
I updated the above snippet so they're both `Ty`s now.

Reply:
Added your suggestion in 8e341449